pkg/clientgocred/cred_test.go

```diff
@@ -0,0 +1,217 @@
+package clientgocred
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"golang.org/x/oauth2"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+type mockTokenSource struct{}
+
+func (*mockTokenSource) Token() (*oauth2.Token, error) {
+	return &oauth2.Token{
+		TokenType:   "bearer",
+		AccessToken: "default_access_token",
+		Expiry:      time.Now().Add(time.Hour),
+	}, nil
+}
+
+func TestExecCredential(t *testing.T) {
+	testCases := []struct {
```

Comment:
I'm pretty uncomfortable with how little of the file-handling code is being exercised (it's all mocked out in every test case).  Golang has pretty reasonable support for creating stand-in filesystems; while we wont' be able to test *real* filesystem behavior, we will at least be able to exercise the "bottom third" of the module's code from the unit tests.

1. https://stackoverflow.com/questions/16742331/how-to-mock-abstract-filesystem-in-go
2. https://godocs.io/testing/fstest#MapFS

You can use mapfs to put more realistic cache entries in the "filesystem" and see how they are handled, as well as to exercise the various locking-related edge cases.

Reply:
Nearly all of the file handling code got removed thanks to the atomic writefile.